### PR TITLE
arena trap and other similar abilities 

### PR DESCRIPTION
**Pull request by Reldnahc**
have applyCheckTrappedAbAttrs() send the pokemon that is being prevented from switching also.

implement conditional arena traps.

implement magnet pull.

correct arena trap to only effect grounded pokemon.

correct shadow tag to only effect non-ghost types.

### CODE REVIEW

File: src/data/ability.ts [hunk 1]
```diff
@@ -2733,7 +2750,7 @@ export function initAbilities() {
     new Ability(Abilities.INTIMIDATE, 3)
       .attr(PostSummonStatChangeAbAttr, BattleStat.ATK, -1, false, true),
     new Ability(Abilities.SHADOW_TAG, 3)
-      .attr(ArenaTrapAbAttr),
+      .attr(ConditionalArenaTrapAbAttr, (opponent: Pokemon) => !opponent.isOfType(Type.GHOST)),
```
Comment (TempsRay) [inline src/data/ability.ts:2753]: The check for ghost type can be applied in ConditionalArenaTrapAbAttr since all ghosts are immune to being trapped

File: src/data/ability.ts [hunk 2]
```diff
@@ -2866,7 +2881,7 @@ export function initAbilities() {
       .attr(PostSummonWeatherChangeAbAttr, WeatherType.SUNNY)
       .attr(PostBiomeChangeWeatherChangeAbAttr, WeatherType.SUNNY),
     new Ability(Abilities.ARENA_TRAP, 3)
-      .attr(ArenaTrapAbAttr)
+      .attr(ConditionalArenaTrapAbAttr, (opponent: Pokemon) => opponent.isGrounded())
```
Comment (TempsRay) [inline src/data/ability.ts:2884]: I believe the `isGrounded` might be a little bit more complicated than just checking a boolean since it needs to account for a lot of factors, but that might be more of a different PR to account for all the intricacies of being grounded

File: src/data/ability.ts (hunk 1 above)
Comment (Reldnahc) [reply src/data/ability.ts:2753]: I didn't know. I will fix this!

**Comment by TempsRay** [comment]
Oh yeah, apparently the ability run away lets you run away (but not switch out) even if arena trap'd or shadow tag'd

**Comment by bennybroseph** [comment]
Closing this due to inactivity. Feel free to make a new request if interest is reignited
